### PR TITLE
#163388131 Add jwt authentication

### PR DESCRIPTION
**Pull request by asirvex**
**What does this PR do?**
- Assign token upon user registration
- Assign token upon user login
- Ensure all endpoints call the authenticate method
- Write tests for the feature

**How should this be manually tested?**
1. Clone the repo using the following command:
`git clone https://github.com/andela/ah-the-phoenix.git`
2. cd into the ah-the-phoenix folder:
`cd ah-the-phoenix`
3. Checkout to the branch `ft-jwt-auth-163388131` with the command:
git checkout `ft-jwt-auth-163388131`
4. Install the requirements by:
`pip install -r requirements.txt`
5. Create a database: 
`create database thephoenix`
6. Export the environment settings as follows:
    `export DATABASE="thephoenix"`
    `export USER="your postgres username "`
    `export HOST="localhost"`
    `export PASSWORD="your postgres password"`
7. Migrate the models to the db:
    `python manage.py migrate`
8. Run the server:
    `python manage.py runserver`
9. From postman:
    1. Create a user by sending a post request to the '/api/users/' endpoint with the data body
    `{`
    `"user": {`
    `"username" :"prefered name",`
    `"email": "your email",`
    `"password": "password"`
    `}`
     You should get a response with a token field which displays the token
     2. Log in by sending a post request to the endpoint '/api/users/login' with the data body
     `{`
    `"user": {`
    `"username" :"prefered name",`
    `"password": "password"`
    `}`
     You should get a response with a token field which displays the token if all goes well

**Any background context you want to provide?**
None

**What are the relevant pivotal tracker stories?**
[#163388131](https://www.pivotaltracker.com/story/show/163388131)

### CODE REVIEW

File: authors/apps/authentication/backends.py [hunk 1]
```diff
@@ -1,10 +1,45 @@
-# import jwt
-#
-# from django.conf import settings
-#
-# from rest_framework import authentication, exceptions
-#
-# from .models import User
+import jwt
```
Comment (dev-jey) [inline authors/apps/authentication/backends.py:1]: Great work Asava. However, you can consider eliminating this space for consistency

Comment (Arusey) [reply authors/apps/authentication/backends.py:1]: Looks good Asava.

File: authors/apps/authentication/backends.py [hunk 2]
```diff
@@ -1,10 +1,45 @@
...
+from .models import User
+
+class JWTAuthentication(authentication.BaseAuthentication):
+    authentication_header_prefix = 'token'
+
+    def authenticate(self, request):
+        """This method is called on every endpoint."""
+        request.user = None
+        auth_header = authentication.get_authorization_header(request).split()
+        auth_header_prefix = self.authentication_header_prefix.lower()
```
Comment (sirghiny) [inline authors/apps/authentication/backends.py]: Is the `.lower()` necessary given its already lowercase in line 10.

File: authors/apps/authentication/models.py [hunk 3]
```diff
@@ -105,6 +105,12 @@ def __str__(self):
         return self.email
 
     @property
+    def token(self):
+        """
+        Allows us to get the token by calling `user.token` instead of
```
Comment (sirghiny) [inline authors/apps/authentication/models.py:111]: Ensure your doc-strings follow pep-8 standards. Refer: https://blog.dolphm.com/pep257-good-python-docstrings-by-example/

File: authors/apps/authentication/backends.py [hunk 4]
```diff
@@ -1,10 +1,45 @@
...
+        if not auth_header:
+            return None
```
Comment (sirghiny) [inline authors/apps/authentication/backends.py:21]: Raise `AuthenticationFailed` exceptions with unique messages for each of these header related mistakes.

**Comment by PaulinhoNdegwa** [comment]
Hi @asirvex. This is a good implementation of jwt authentication. You've done a good job at raising errors and testing them.

File: authors/apps/authentication/backends.py [hunk 5]
```diff
@@ -1,10 +1,55 @@
...
+    authentication_header_prefix = 'token'
+
+    def authenticate(self, request):
+        """Checks authorization on every request."""
+
+        request.user = None
+        auth_header = authentication.get_authorization_header(request).split()
+        
+        if not auth_header:
+            return None
```
Comment (sirghiny) [inline authors/apps/authentication/backends.py:21]: Raise an `AuthenticationFailed` exception for missing headers too.

Comment (asirvex) [reply authors/apps/authentication/backends.py:21]: The method is called on each endpoint. If I raise an error on missing error, an error will be returned on the login and signup endpoints too which dont require jwt authentication

**Pull request merged**